scroll-tech/zkevm-circuits
# TOWARDS E2E CONTRACT CALL PROVING!

The aim of this PR is to prove e2e contract call.   `set_value then retrieve` in `integration-tests/contracts/greeter/Greeter.sol` is planned.

Things done:

0. add refund to geth trace
1. add **basic happy path** bus mapping for sstore and sload,  and tests passed
2. some bugs fixed: sstore sload should use SWC transtion `Delta` rather than `To`
3. some adhoc fixes: (1) in opcode bus mapping, we should access storage **before** current step to get prev_value, currently i use `unsafe { ptr = steps.as_ptr(); *(ptr - 1).storage`, we should rewrite it later (2) underflow occured in sstore circuit, i temporarily add a constant to all intimidate values and then sub it finally... 
4. add lookup id log for debugging
5. generate contract call data in integration-tests
6. finish is_zero opcode
7. connect some opcodes between evm and busmapping in witness.ts, including calldatasize
8. merge SHR (apply some fixes), merge bus-mapping CALLDATACOPY
9. use non EIP1559 ethers-rs
10. better degree calculation of evm_circuit


TODOs:

1. implement (non-adhoc) bus mapping for calldataload
2. traces should have correct prev_value / commited_value / etc for sload sstore 



How to test:

0. install docker locally
1. cd integration-tests
4. ./run.sh --steps "setup"
5. ./run.sh --steps "gendata"
6. RUST_LOG=circuits,bus_mapping=trace cargo test --profile release --test circuits --features circuits test_evm_circuit_block_greeter_calls -- --nocapture

then you can see

```
test test_evm_circuit_block_greeter_calls ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 6 filtered out; finished in 15.48s
```



More thoughts:

1. the storage of opcodes in trace is O(N**2). We should optimize it later...
2. for every opcode, we should implement. bus-mapping tests (even happy path only) to detect real world problems.. like calldataload / sstore / shr
3. In calldataload, is it reasonable to use many & dynamic number of lookups..?
4. how to debug complex lookups? `Lookup { name: "LOOKUP: \"RW table\"", lookup_index: 137, location: InRegion { region: Region { index: 5, name: "Execution step" }, offset: 272 } }`, just print offset and lookup_index "assign_exec_step 272 ExecStep { call_index: 0, rw_indices: [(Stack, 22), (Stack, 23)], execution_state: CALLDATALOAD, rw_counter: 76, program_counter: 28, stack_pointer: 1023, gas_left: 478730, gas_cost: 3, memory_size: 96, state_write_counter: 2, opcode: Some(CALLDATALOAD), aux_data: None }" AND "LOOKUP TABLE 137 <=> "RW table" 1"
